File: course1_algorithmic_toolbox/week4_divide_and_conquer/1_binary_search/binary_search.py
# Uses python3
import sys
import math

def binary_search(arr, key):
    min, max = 0, (len(arr) - 1)
    while(min <= max):
        mid = math.floor((min+max)/2)
        if (arr[mid] == key): return mid
        if(arr[mid] > key): max = mid - 1
        if(arr[mid] < key): min = mid + 1
    return -1

def linear_search(a, x):
    for i in range(len(a)):
        if a[i] == x:
            return i
    return -1

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n = data[0]
    m = data[n + 1]
    a = data[1 : n + 1]
    for x in data[n + 2:]:
        print(binary_search(a, x), end = ' ')

File: course1_algorithmic_toolbox/week4_divide_and_conquer/1_binary_search/binary_search.js

const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    terminal: false
});

process.stdin.setEncoding('utf8');
rl.once('line', line => {
    const arr = line.toString().split(' ').slice(1).map(Number);

    rl.once('line', line => {
        const keys = line.toString().split(' ').slice(1).map(Number);
        const result = [];
        for (let key of keys) {
            result.push(binarySearch(arr, key));
        }
        
        const res = result.join(' ');
        const maxLength = 50000;

        for (let i = 0; i < res.length; i += maxLength) {
            process.stdout.write(res.slice(i, i + maxLength));
        }

        process.stdout.write('\n');
        process.exit();
    })
});

function binarySearch(arr = [], key) {
    // check the middle element
    let min = 0
    let max = (arr.length - 1)
    while(min <= max){
        const mid = Math.floor((min+max)/2)
        if (arr[mid] === key) return mid;
        if(arr[mid] > key) max = mid - 1
        if(arr[mid] < key) min = mid + 1
    }
    return -1
}

module.exports = binarySearch;
